store/tikv/lock_resolver.go

```diff
@@ -146,6 +149,79 @@ func (lr *LockResolver) getResolved(txnID uint64) (TxnStatus, bool) {
 	return s, ok
 }
 
+// BatchResolveLocks resolve locks in a batch
+func (lr *LockResolver) BatchResolveLocks(bo *Backoffer, locks []*Lock, loc RegionVerID) (bool, error) {
+	if len(locks) == 0 {
+		return true, nil
+	}
+
+	lockResolverCounter.WithLabelValues("resolve").Inc()
+
+	var expiredLocks []*Lock
+	for _, l := range locks {
+		if lr.store.GetOracle().IsExpired(l.TxnID, l.TTL) {
+			lockResolverCounter.WithLabelValues("expired").Inc()
+			expiredLocks = append(expiredLocks, l)
+		} else {
+			lockResolverCounter.WithLabelValues("not_expired").Inc()
+		}
+	}
+	if len(expiredLocks) == 0 {
+		log.Error("This should not occur!")
+		return true, nil
+	}
+
+	lr.cntTxnStatus = 0
+	lr.costTxnStatus = 0
+	txnInfos := make(map[uint64]uint64)
+	count := 0
+	for _, l := range expiredLocks {
+		count = count + 1
+		status, err := lr.getTxnStatus(bo, l.TxnID, l.Primary)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		txnInfos[l.TxnID] = uint64(status)
+	}
+	log.Infof("BatchResolveLocks: it takes %v cost to lookup %v txn status from %v txn status", lr.costTxnStatus, lr.cntTxnStatus, count)
+
+	var listTxnInfos []*kvrpcpb.TxnInfo
+	for txnID, status := range txnInfos {
+		listTxnInfos = append(listTxnInfos, &kvrpcpb.TxnInfo{
+			Txn:    txnID,
+			Status: status,
+		})
+	}
+
+	req := &tikvrpc.Request{
+		Type: tikvrpc.CmdResolveLock,
+		ResolveLock: &kvrpcpb.ResolveLockRequest{
+			TxnInfos: listTxnInfos,
+		},
+	}
+	startTime := time.Now()
+	resp, err := lr.store.SendReq(bo, req, loc, readTimeoutShort)
+	log.Infof("BatchResolveLocks: it takes %v cost to resolve %v locks in a batch.", time.Since(startTime), len(expiredLocks))
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	regionErr, err := resp.GetRegionError()
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	if regionErr != nil {
+		err = bo.Backoff(BoRegionMiss, errors.New(regionErr.String()))
```

Comment:
What is the backoff's purpose here?

Reply:
maybe we'd better keep it here, this `backoff` is used for `region error`. If we remove it, and rely on the `backoff` outside, it may not work as we expect in the future.

Reply:
Got it.